lib/src/session.dart

```diff
@@ -0,0 +1,84 @@
+import 'dart:convert';
+
+import 'package:clock/clock.dart';
+import 'package:file/file.dart';
+import 'package:path/path.dart' as p;
+
+import 'constants.dart';
+
+class Session {
+  final Directory homeDirectory;
+  final FileSystem fs;
+  final File _sessionFile;
+  late int _sessionId;
```

Comment:
avoid `late` except in tests when they are initialized by `setUp()`. One trick to allow a field to be non-nullable when it has to be calculated from a function call is to create a factory that makes the necessary calls (here calls to `.readAsStringSync()` and `jsonDecode`, then returns the result of a call to a private constructor that then initializes the non-nullable fields (e.g. `Session._({this._sessionId, this._lastPing, etc...})`).

Reply:
This was a great idea, thank you!